automata/pda/pda.py

```diff
@@ -24,6 +52,325 @@ class PDA(Automaton, metaclass=abc.ABCMeta):
+            [start_color, end_color], space="srgb"
+        )
+
+        # find all transitions in the finite state machine with traversal.
+        for transition_index, (from_state, to_state) in enumerate(input_path, start=1):
+            color = interpolation(transition_index / len(input_path))
+
+            symbol = self._get_symbol_configuration(from_state, to_state)
+            label = self._get_edge_name(*symbol)
+
+            edge_drawn[from_state.state, to_state.state, symbol] = True
+            graph.add_edge(
+                self._get_state_name(from_state.state),
+                self._get_state_name(to_state.state),
+                label=f"<{label} <b>[<i>#{transition_index}</i>]</b>>",
+                arrowsize=arrow_size,
+                fontsize=font_size,
+                color=color.to_string(hex=True),
+                penwidth="2.5",
+            )
+
+        return graph
+
+    def _add_transitions_diagram(
+        self,
+        graph: pgv.AGraph,
+        is_edge_drawn: EdgeDrawnDictT,
+        arrow_size: str,
+        font_size: str,
+    ) -> pgv.AGraph:
+        """
+        Add transitions to between states
+        """
+
+        edge_labels = defaultdict(list)
+        for from_state, to_state, symbol in self.iter_transitions():
+            if is_edge_drawn[from_state, to_state, symbol]:
+                continue
+
+            from_node = self._get_state_name(from_state)
+            to_node = self._get_state_name(to_state)
+            label = self._get_edge_name(*symbol)
+            edge_labels[from_node, to_node].append(label)
+
+        for (from_node, to_node), labels in edge_labels.items():
+            graph.add_edge(
+                from_node,
+                to_node,
+                label="\n".join(sorted(labels)),
+                arrowsize=arrow_size,
+                fontsize=font_size,
+            )
+
+        return graph
+
+    def _add_states_diagram(
+        self,
+        graph: pgv.AGraph,
+        arrow_size: str,
+        font_size: str,
+    ) -> pgv.AGraph:
+        """
+        Add all the states of the PDA
+        """
+
+        # create unique id to avoid colliding with other states
+        null_node = create_unique_random_id()
+
+        graph.add_node(
+            null_node,
+            label="",
+            tooltip=".",
+            shape="point",
+            fontsize=font_size,
+        )
+        initial_node = self._get_state_name(self.initial_state)
+        graph.add_edge(
+            null_node,
+            initial_node,
+            tooltip="->" + initial_node,
+            arrowsize=arrow_size,
+        )
+
+        nonfinal_states = map(self._get_state_name, self.states - self.final_states)
+        final_states = map(self._get_state_name, self.final_states)
+        graph.add_nodes_from(nonfinal_states, shape="circle", fontsize=font_size)
+        graph.add_nodes_from(final_states, shape="doublecircle", fontsize=font_size)
+
+        return graph
```

Comment:
Aren't all of these changes made to the object in-place? If so, there's no need to return anything. Same goes for all of the other helper functions that are returning the graph.